### DISCUSSION TITLE
device restart from script?

### DISCUSSION THREAD

**Discussion by jig432** [arendst/Tasmota#10551] [Q&A] [answered]
I just had a tasmota (esp32) disconnect from my mqtt server and constantly retry and never reconnect.  A manual  restart fixed it.  I'm running a tasmota image with scripts enabled.  I'd like to automatically restart the device after being disconnected for "too long". I plan to use the ">S" section to count the number of times I detect mqtt disconnected in sequence and trigger a restart after N times.  However, I don't see a script command to do a restart.  So, 2 questions:
1. Is there a way to have a script cause a restart? If not, in the source I see use of EspRestart() in the restart console command. if I added a script command to do a restart, Is that the right function to invoke to cause a restart?
2. Is there some other/better way to detect mqtt disconnect and do something to cause some form of stronger retry? Maybe, restart wifi?

**Comment by gemu2015** [comment, marked as answer]
1. ->restart 1
2. not to my knowledge

**Reply by jig432** [reply]
1. Oops.... tried to find how to use console commands and just missed it. Thanks.

**Reply by tauuu** [reply]
Can you share reset sript or rule? I have same poblem with sonoffs disconnections about one per month...

**Reply by gemu2015** [reply]
this script resets every night at 3 am

\>D
h=0
\>S
h=hours
if chg[h]>0
and h==3
then
=>restart 1
endif

**Reply by Wuifi** [reply]
I tried the script mentioned by gemu2015 on two different devices. It restarts the ESP, but unfortunately after the first restart the script is disabled.
Does anybody experienced the same and has a solution?
Regards
Wolfgang

**Comment by barbudor** [comment]
There are rules trigger to detect wifi or mqtt disconnections from which you can call 'restart 1'
Triggers are listed in the Rules doc

**Reply by barbudor** [reply]
However it should be better to get the problem from the beginning and solve it instead of working around it

**Reply by sfromis** [reply]
To manage disconnection issues in a comprehensive way, I'd not recommend trying to detect outage, but instead have the backend publish a regular "heartbeat" pulse, and then clients can take action if not getting that. One simple type of heartbeat would be to power on a relay where PulseTime is configured. If the timer pops, something is rotten. It could be a "virtual relay" on an unused gpio, or then maybe use a RuleTimer instead.
